src/Npgsql/Internal/Converters/RangeConverter.cs

```diff
@@ -0,0 +1,189 @@
+            return NpgsqlRange<T>.Empty;
+
+        var lowerBound = default(T);
+        var upperBound = default(T);
+
+        if ((flags & RangeFlags.LowerBoundInfinite) == 0)
+        {
+            if (reader.ShouldBuffer(sizeof(int)))
+                await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+            var length = reader.ReadInt32();
+
+            // Note that we leave the CLR default for nulls
+            if (length != -1)
+            {
+                // Set size before calling ShouldBuffer (it needs to be able to resolve an upper bound requirement)
+                reader.Current.Size = length;
+                if (reader.ShouldBuffer(_subtypeBufferReadRequirements))
+                    await reader.BufferData(async, _subtypeBufferReadRequirements, cancellationToken).ConfigureAwait(false);
+
+                lowerBound = async
+                    ? await _subtypeConverter.ReadAsync(reader, cancellationToken)
+                    // ReSharper disable once MethodHasAsyncOverloadWithCancellation
+                    : _subtypeConverter.Read(reader);
+            }
+        }
+
+        if ((flags & RangeFlags.UpperBoundInfinite) == 0)
+        {
+            if (reader.ShouldBuffer(sizeof(int)))
+                await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+            var length = reader.ReadInt32();
+
+            // Note that we leave the CLR default for nulls
+            if (length != -1)
+            {
+                // Set size before calling ShouldBuffer (it needs to be able to resolve an upper bound requirement)
+                reader.Current.Size = length;
+                if (reader.ShouldBuffer(_subtypeBufferReadRequirements))
+                    await reader.BufferData(async, _subtypeBufferReadRequirements, cancellationToken).ConfigureAwait(false);
+                upperBound = async
+                    ? await _subtypeConverter.ReadAsync(reader, cancellationToken)
+                    // ReSharper disable once MethodHasAsyncOverloadWithCancellation
+                    : _subtypeConverter.Read(reader);
+            }
+        }
+
+        return new NpgsqlRange<T>(lowerBound, upperBound, flags);
+    }
+
+    public override Size GetSize(SizeContext context, NpgsqlRange<T> value, ref object? writeState)
+    {
+        var totalSize = Size.Create(1); // Flags
+
+        if (!value.IsEmpty)
+        {
+            var rangeWriteState = new WriteState();
+
+            if (!value.LowerBoundInfinite)
+            {
+                var size = Size.Create(4); // Length
+
+                if (!_subtypeConverter.IsDbNull(value.LowerBound))
+                    size = size.Combine(_subtypeConverter.GetSize(context, value.LowerBound, ref rangeWriteState.LowerBoundWriteState));
+                rangeWriteState.LowerBoundSize = size.Value - 4;
+                totalSize = totalSize.Combine(size);
```

Comment:
Maybe just have a + operator for Size?

Reply:
As a Size is still kind of a commutative semigroup we technically could.. but I'm usually not a huge fan of hiding this kind of logic under an operator that we all think we know.

Reply:
Up to you

Reply:
Leave it for now